### DISCUSSION TITLE
GraphQL `codegen` not finding tables (only generating basic scalars etc)

### DISCUSSION THREAD

**Discussion by prescience-data** [supabase/supabase#6262] [Questions] [answered]
Most of my tables are restricted to `authenticated` or higher, so it doesn't appear that the `codegen` `getIntrospectionQuery()` request is able to introspect even when passing it a `service_role` key.

Interestingly, even with
```pgsql
GRANT SELECT ON TABLE countries, regions, timezones TO anon;
```
still doesn't appear to be able to gain visibility, so I'm leaning toward a possible misunderstanding / misconfiguration on my side. 

Any hints would be appreciated. 

### Additional Context

This is a new project created yesterday and `pg_graphql` extension is enabled.

```pgsql

DROP SCHEMA IF EXISTS public CASCADE;
CREATE SCHEMA public;
-- AlterPrivileges
ALTER DEFAULT PRIVILEGES REVOKE EXECUTE ON FUNCTIONS FROM anon;
ALTER DEFAULT PRIVILEGES REVOKE ALL ON TABLES FROM anon;
-- GrantUsage
GRANT USAGE ON SCHEMA public TO postgres, supabase_admin, service_role, authenticated, anon;
-- GrantAll
GRANT ALL ON ALL TABLES IN SCHEMA public TO postgres, supabase_admin, service_role;
-- GrantSelect
GRANT SELECT ON TABLE countries, regions, timezones, platforms, proxies TO anon;
-- GraphQL
COMMENT ON SCHEMA public IS e'@graphql({"inflect_names": true})';

```

Output from `codegen:fetch`
```gql

scalar BigInt

"""
Boolean expression comparing fields on type "BigInt"
"""
input BigIntFilter {
  eq: BigInt
  gt: BigInt
  gte: BigInt
  lt: BigInt
  lte: BigInt
  neq: BigInt
}

"""
Boolean expression comparing fields on type "Boolean"
"""
input BooleanFilter {
  eq: Boolean
  gt: Boolean
  gte: Boolean
  lt: Boolean
  lte: Boolean
  neq: Boolean
}

scalar Cursor

scalar Date

"""
Boolean expression comparing fields on type "Date"
"""
input DateFilter {
  eq: Date
  gt: Date
  gte: Date
  lt: Date
  lte: Date
  neq: Date
}

scalar Datetime

"""
Boolean expression comparing fields on type "Datetime"
"""
input DatetimeFilter {
  eq: Datetime
  gt: Datetime
  gte: Datetime
  lt: Datetime
  lte: Datetime
  neq: Datetime
}

"""
Boolean expression comparing fields on type "Float"
"""
input FloatFilter {
  eq: Float
  gt: Float
  gte: Float
  lt: Float
  lte: Float
  neq: Float
}

"""
Boolean expression comparing fields on type "Int"
"""
input IntFilter {
  eq: Int
  gt: Int
  gte: Int
  lt: Int
  lte: Int
  neq: Int
}

scalar JSON

"""
Boolean expression comparing fields on type "JSON"
"""
input JSONFilter {
  eq: JSON
  neq: JSON
}

"""Defines a per-field sorting order"""
enum OrderByDirection {
  AscNullsFirst
  AscNullsLast
  DescNullsFirst
  DescNullsLast
}

type PageInfo {
  endCursor: String
  hasNextPage: Boolean!
  hasPreviousPage: Boolean!
  startCursor: String
}

"""The root type for querying data"""
type Query {
  """UTC Datetime from server"""
  heartbeat: Datetime!
}

"""
Boolean expression comparing fields on type "String"
"""
input StringFilter {
  eq: String
  gt: String
  gte: String
  lt: String
  lte: String
  neq: String
}

scalar Time

"""
Boolean expression comparing fields on type "Time"
"""
input TimeFilter {
  eq: Time
  gt: Time
  gte: Time
  lt: Time
  lte: Time
  neq: Time
}

scalar UUID

"""
Boolean expression comparing fields on type "UUID"
"""
input UUIDFilter {
  eq: UUID
  neq: UUID
}

```

### Additional Debugging Steps so far

1. Just created a totally fresh project and created a single table via Dashboard on `public` schema with no RLS.
2. Enabled `pg_graphql`.
3. Attempted to introspect schema, and received no results.

At this point I'm now extremely certain that I must doing something wrong here, but not sure what it could be given (what seems to be) a 1:1 mirror of the Supabase graphql instructions / sample code on a fresh project.

**Comment by prescience-data** [comment, marked as answer]
Ok finally found the issue at https://supabase.com/docs/guides/api#graphql-api-1

> _While pg_graphql is in beta, we have disabled GraphQL auto-refreshing. If you want to refresh your GraphQL introspection, call select graphql.rebuild_schema(); to rebuild your GraphQL schema._

This works when running manually from the Dashboard, however it seems the `postgres` user which is used to connect to the database directly via password does not have permissions to `TRUNCATE` or `DELETE` the tables in `graphql.rebuild_schema()`

```pgsql
CREATE FUNCTION rebuild_schema
(
) RETURNS void
  LANGUAGE plpgsql
AS
$$
begin
    truncate table graphql._field;
    delete from graphql._type;
    refresh materialized view graphql.entity with data;
    refresh materialized view graphql.entity_column with data;
    refresh materialized view graphql.entity_unique_columns with data;
    refresh materialized view graphql.relationship with data;
    perform graphql.rebuild_types();
    perform graphql.rebuild_fields();
end;
$$;
ALTER FUNCTION rebuild_schema() OWNER TO supabase_admin;
GRANT EXECUTE ON FUNCTION rebuild_schema() TO postgres;
GRANT EXECUTE ON FUNCTION rebuild_schema() TO anon;
GRANT EXECUTE ON FUNCTION rebuild_schema() TO authenticated;
GRANT EXECUTE ON FUNCTION rebuild_schema() TO service_role;


```

...which seems strange given the `EXECUTE` grants to `postgres`, `anon`, and `authenticated` (given this function will fail due to lack of permissions).

Further, the `postgres` user does not have permissions to grant itself those methods, so it appears the only way to refresh is via the Dasboard?

Related bug report with interim fix for `postgres` user issue: https://github.com/supabase/supabase/issues/6271

**Reply by olirice** [reply]
cross posting for visibility

> This works when running manually from the Dashboard, however it seems the postgres user which is used to connect to the database directly via password does not have permissions to TRUNCATE or DELETE the tables in graphql.rebuild_schema()

we're tracking that issue at https://github.com/supabase/supabase/issues/6271
